### PR TITLE
Platform: Mac App Store support

### PR DESCRIPTION
**Pull request by sethlu**
- MAS platform support
- Entitlements (optional) are supported to sign apps, not just with a certificate
- Made it easier to sign for MAS distribution, may fallback to default entitlements file

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -18,7 +18,8 @@ var supportedPlatforms = {
   // Maps to module ID for each platform (lazy-required if used)
   darwin: './mac',
   linux: './linux',
-  win32: './win32'
+  win32: './win32',
+  mas: './mac' // it is almost the same as the mac version
```
Comment (malept) [inline index.js]: This map should be sorted alphabetically.

File: index.js [hunk 2]
```diff
@@ -68,6 +69,7 @@ function createSeries (opts, archs, platforms) {
     platforms.forEach(function (platform) {
       // Electron does not have 32-bit releases for Mac OS X, so skip that combination
       if (platform === 'darwin' && arch === 'ia32') return
+      if (platform === 'mas' && arch === 'ia32') return
```
Comment (malept) [inline index.js]: This line and the previous line should be merged together.

File: readme.md [hunk 3]
```diff
@@ -177,6 +177,8 @@ If the file extension is omitted, it is auto-completed to the correct extension
 
 `sign` - *String*
 
+`sign-entitlements` - *String*
+
```
Comment (malept) [inline readme.md]: Please add a description of the parameter here.

File: test/multitarget.js [hunk 4]
```diff
@@ -38,8 +38,8 @@ test('all test', function (t) {
     function (cb) {
       packager(opts, cb)
     }, function (finalPaths, cb) {
-      // Windows skips packaging for OS X, and OS X only has 64-bit releases
-      t.equal(finalPaths.length, process.platform === 'win32' ? 4 : 5,
+      // Windows skips packaging for OS X, and OS X only has 64-bit releases and mas
```
Comment (malept) [inline test/multitarget.js]:
```javascript
// Windows skips packaging for OS X (darwin + mas), and OS X only has 64-bit releases
```

File: mac.js [hunk 5]
```diff
@@ -100,9 +100,29 @@ module.exports = {
...
+            operations.push(function (cb) {
+              child.exec('codesign --deep --force --sign "' + opts.sign + '" --entitlements "' + inheritEntitlementsFile + '" "' + entry + '"', cb)
+            })
+          })
+          operations.push(function (cb) {
+            child.exec('codesign --force --sign "' + opts.sign + '" --entitlements "' + (opts['sign-entitlements'] || path.join(__dirname, 'mas-default.entitlements')) + '" "' + finalAppPath + '"', cb) // may fallback to default entitlements
+          })
+        } else {
+          operations.push(function (cb) {
+            child.exec('codesign --deep --force --sign "' + opts.sign + '" "' + finalAppPath + '"', cb)
```
Comment (malept) [inline mac.js]: It feels like the three `codesign` calls could be DRY'd up a bit.

**Comment by malept** [comment]
Please update the rest of the "darwin platform only" references to be "darwin/mas platforms only".

**Comment by malept** [comment]
This would solve #163 (or at least part of it). I'm going to wait on @maxogden's opinion on this before merging, as he had some concerns in the linked issue.

If someone with a Mac could test this, it would be appreciated.

**Comment by maxogden** [comment]
Here's my opinion:

- Looks pretty solid + straightforward, however:
- I think there needs to be some mention *somewhere* in the readme that `mas` means "Mac App Store" just to help user friendliness
- The fact that we're adding a bunch of code signing specific code with no tests makes me really nervous. I would *really* prefer it at this point if all the codesigning stuff got moved into a separate standalone module that we could depend on. This kind of patch is a big red flag to me

**Comment by malept** [comment]
> I think there needs to be some mention somewhere in the readme that mas means "Mac App Store" just to help user friendliness

:+1: - also, linking to the [Electron docs](http://electron.atom.io/docs/v0.36.0/tutorial/mac-app-store-submission-guide/) would be helpful.

> The fact that we're adding a bunch of code signing specific code with no tests makes me really nervous.

Yeah, me too. The problem I have is that I don't know what a reasonable way to programatically test this looks like.

> I would _really_ prefer it at this point if all the codesigning stuff got moved into a separate standalone module that we could depend on. This kind of patch is a big red flag to me

:+1:

**Comment by maxogden** [comment]
> The problem I have is that I don't know what a reasonable way to programatically test this looks like.

Agreed, it's a hard one. I think in place of proper tests the next best thing would be to put all the signing code into a new module, e.g. `electron-sign`, and then at least it can have a dedicated issue tracker, version number, external API, and maintainers. 

Is anyone up for writing + maintaining a `electron-sign` module?

**Comment by sethlu** [comment]
Putting the codesign into a separated module for building and testing could be a nice track, as testing online proves hard for signing with certificates, needing manual help after all. I'll see if I could start a module integrating the OS X codesign command later.
Here's a list of things I guess might be confusing for those packing the mac apps:
- The certificates for packing an app for distributing outside/in the Mac App Store are different.
- Distributing outside the Mac App Store doesn't necessarily need the entitlements, as they are for sandboxing as the main purpose (to my perspective).
- From my reading online, I found code signing each file in the project is recommended(?) by Apple as Xcode does so, with --deep ignored.
- For distribution within the Mac App Store, an installer package should be required for uploading the app to iTunesConnect for reviewing (so I may put up a electron-sign with installer)

**Comment by sethlu** [comment]
Here's what I've got from today: https://github.com/sethlu/electron-sign
Tbh, only those with the dev program could have the code tested.

**Comment by maxogden** [comment]
@sethlu Wow!! Amazing work. That looks really good so far. I looked at the code and didn't see any obvious problems. It's great that you have a test suite and `standard` hooked up too :)

When you are ready, do you want to make a PR to electron-packager that integrates with electron-sign? I can also add you to the electron-packager collaborators so you can help maintain the integration in the future.

**Comment by cjb** [comment]
(Maybe electron-mac-sign or something, since Windows Electron apps also require signing?)

**Comment by maxogden** [comment]
@cjb is there any existing code for that? if if's not too hairy maybe it could live in `electron-sign` too?

**Comment by sethlu** [comment]
@maxogden @cjb I've renamed the project to `electron-osx-sign` and it should avoid some issues with understanding from simply reading the command, as it is OS X only now. Seems that there is as well some sorts of code signing with Windows, but I wouldn't get into that too much now for Electron as distribution to OS X's a bit more complex in following the steps, and it's quite needing.
An updated repo: https://github.com/sethlu/electron-osx-sign
I'll see how I could have my `electron-packager` fork updated with code-sign and make several tests before doing anything further.

**Comment by maxogden** [comment]
@sethlu excellent, sounds good, thanks!

**Comment by malept** [comment]
Regarding signing packages on Windows: #32

**Comment by sethlu** [comment]
I'm not quite sure about the code signing for the Windows platform as I do most of my work on my Mac.
I haven't replaced the testing procedures for codesign around test/mac.js:181, not sure if to remove it as the code signing's been moved to a separate module.

File: package.json [hunk 6]
```diff
@@ -27,7 +27,8 @@
     "plist": "^1.1.0",
     "rcedit": "^0.3.0",
     "rimraf": "^2.3.2",
-    "run-series": "^1.1.1"
+    "run-series": "^1.1.1",
+    "electron-osx-sign": "^0.1.4"
```
Comment (malept) [inline package.json]: Please keep the dependency list alphabetized.

**Comment by maxogden** [comment]
protip, if you use `npm install --save some-dependency` then `npm` will automatically add it to the dependencies in package.json and alphabetize it for you :)

**Comment by sethlu** [comment]
@maxogden Wow, thanks! I'll have it a try the next time I need to have the package description sorted.
On the travis build, I'm thinking about pulling the timeout longer for testing as it frequently does.

**Comment by sethlu** [comment]
Oops

**Comment by malept** [comment]
Here is what I think the status of this PR is:

* [ ] Rebase on master
* [ ] Needs docs updated (*[ref](https://github.com/maxogden/electron-packager/pull/223#issuecomment-166159352)*) - I can help with that, if you like.
* [ ] Reduce the amount of duplicated code between `test/mac.js` and `test/mas.js`. A good chunk of that can be put into a separate common module (`test/common-mac.js`?).
* [ ] Squash commits

Once these are all done, I think we're good to merge.

**Comment by marbemac** [comment]
Would love to see this! Was the decision in the end to include mas functionality in this package, or a separate one?

**Comment by malept** [comment]
@marbemac in this package, with signing in (one or more) separate modules.

**Comment by sethlu** [comment]
Hi, I've merged some pull requests for the signing module, as previously some parts within the framework were not covered for `codesign`. With further testing on this additional module, I could start fine-tune some updates for the MAS distribution packing.

**Comment by sethlu** [comment]
Hi, I've just found my issue reported to `electron` mentioned in the documentation as a known bug here https://github.com/atom/electron/blob/master/docs/tutorial/mac-app-store-submission-guide.md. Thus I'll currently recommend MAS builds before 0.36.0.
I'll be posting some updates later today, which may include fixes for https://github.com/sethlu/electron-osx-sign/issues/4 & https://github.com/maxogden/electron-packager/issues/261 that will keep a consistency in the renaming of helper applications.

**Comment by sethlu** [comment]
So far I've done the following:
- Rebased on master
- Squashed `test/mac.js` and `test/mas.js` to a common and generalized testing script `test/mac.js`, which will be called inside `test/darwin.js` and `test/mas.js` with different `baseOpts`.

File: test/basic.js [hunk 7]
```diff
@@ -49,7 +49,7 @@ function createDefaultsTest (combination) {
         resourcesPath = path.join(finalPath, util.generateResourcesPath(opts))
         fs.stat(path.join(finalPath, generateNamePath(opts)), cb)
       }, function (stats, cb) {
-        if (opts.platform === 'darwin') {
+        if (opts.platform === 'darwin' || opts.platform === 'mas') {
```
Comment (malept) [inline test/basic.js]: I keep seeing variations on this idiom in the code and part of me wants to refactor it to

```javascript
if (platform_is_mac(opts.platform)) {
```

File: index.js [hunk 8]
```diff
@@ -95,7 +96,7 @@ function createSeries (opts, archs, platforms) {
   archs.forEach(function (arch) {
     platforms.forEach(function (platform) {
       // Electron does not have 32-bit releases for Mac OS X, so skip that combination
-      if (platform === 'darwin' && arch === 'ia32') return
+      if ((platform === 'darwin' || platform === 'mas') && arch === 'ia32') return
```
Comment (malept) [inline index.js]: Use `util.isPlatformMac`

File: index.js [hunk 9]
```diff
@@ -161,11 +162,11 @@ function createSeries (opts, archs, platforms) {
           })
         }
 
-        if (combination.platform === 'darwin') {
+        if (combination.platform === 'darwin' || combination.platform === 'mas') {
```
Comment (malept) [inline index.js]: Use `util.isPlatformMac`

**Comment by malept** [comment]
Could you add tests for your fixes to #261 (asserting that the files are named correctly)?

File: index.js (hunk 9 above)
Comment (sethlu) [inline index.js]: As `util.js` is only included during testing. I'll write another isPlatformMac in `index.js` instead of requiring `test/util`?

Comment (malept) [inline index.js]: Yeah, that's better. I should have looked a bit more closely at where it was defined.

**Comment by d-schloss-mvg** [comment]
I dont know if this is the correct place, but using Umlauts or special characters in the name of the application causes these to appear in the bundle-identifiers of the helpers and thus signing is failing. (Signing itself works, but verifying yields a broken signature)

**Comment by sethlu** [comment]
@d-schloss-mvg I think this is an issue with `electron-osx-sign`.
I've already added in the support for special characters in binaries and other paths.

**Comment by d-schloss-mvg** [comment]
The problem are the special characters in the bundle-identifier. Even using the command line tool codesign directly - without any node stuff - fails (so it has nothing to do with electron-osx-sign, right?) if there are these special characters in the bundle-identifier of the helpers.

See [Apple Documentation](https://developer.apple.com/library/mac/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-102070)
Quote: "The bundle ID string must be a uniform type identifier (UTI) that contains only alphanumeric (A-Z,a-z,0-9), hyphen (-), and period (.) characters"
non alphanumeric characters like "ä,ö,ü,é" are not filtered out right now.

**Comment by sethlu** [comment]
@d-schloss-mvg I'll have a test with bundle ids with special characters. Sorry I thought this is about the signing; it now appears to be with the plists that are modified.
Anyway, the signing of apps with special characters in names should now be fixed.

**Comment by d-schloss-mvg** [comment]
Yeah, its definitley the plists. If you just (manually) edit the bundle identifiers in the plists, codesigning works again.

**Comment by sethlu** [comment]
@d-schloss-mvg Great. I guess replacing those special characters with standard characters, like mapping `ä` to `a` should be quite pleasant. Or could replace/remove all of them.
I'm not sure which of the two may be better?

**Comment by sethlu** [comment]
@d-schloss-mvg In my latest commit, all special characters will be removed for the CFBundleIdentifiers in all plists.

**Comment by d-schloss-mvg** [comment]
Wow, thanks for the fast solution!

**Comment by malept** [comment]
@sethlu could you add a test for eb2e4ff (if you aren't doing so already)?

**Comment by malept** [comment]
I've made some documentation changes from a [fork of this PR](https://github.com/malept/electron-packager/tree/mac-app-store-docs#readme), which hopefully clarifies what MAS is and perhaps the difference between `mas` and `darwin`.

@sethlu if the doc changes look reasonable, feel free to cherry-pick the commit into this PR.

**Comment by malept** [comment]
Once we resolve the documentation situation and the commits are squashed appropriately (see [the end of the new section of the contributing docs](https://github.com/maxogden/electron-packager/blob/master/CONTRIBUTING.md#filing-pull-requests) for for more details), I will happily merge this PR.

**Comment by sethlu** [comment]
@malept just rebased on `electron-packager/master`, not sure if it's correct.
I'll still have to amend a few usage/readme notes before eventually closing this.

**Comment by malept** [comment]
Er, don't squash my readme commit.

**Comment by malept** [comment]
@sethlu here's a better squash/rebase (took me two tries to get it in this state): https://github.com/maxogden/electron-packager/compare/master...malept:mas-squashed

What is left in terms of readme and usage changes?

**Comment by kfranqueiro** [comment]
Do you think it's worth noting somewhere/somehow that `--platform=all` is now going to end up including `mas`?  (i.e. if people want to continue building for the targets that `all` used to mean, they'd have to use `--platform=darwin,linux,win32` instead; not sure if this should also mean a major version bump)

**Comment by malept** [comment]
> Do you think it's worth noting somewhere/somehow that `--platform=all` is now going to end up including `mas`?

Yes, this is actually a good impetus for creating a `NEWS.md` in the spirit of http://keepachangelog.com/ (after this PR is merged). 

> not sure if this should also mean a major version bump

I'm not sure either. At minimum, we need to clarify that `--all` and `--platform=all` means that all supported platforms are going to be built (and similar for `--arch=all`, I don't want to have the same conversation whenever it is we're unblocked on the arm7 PR). I would like to do this after this PR is merged.

**Comment by malept** [comment]
Sorry, misclicked.

**Comment by sethlu** [comment]
@malept Just added in the rest of the documentation, and separated your commit for the basic Mac App Store docs in this PR.
I guess the `--platform=all` option might now mean to add platform `mas` into what were previously included, the `darwin`, `linux` and `win32`. As the version check for `mas` is not currently included in the script, an error message will be thrown when someone wants to have `--platform=all` with Electron earlier than `v0.34.0`. Otherwise, we could have `mas` separated, remaining the `all` option as it is now; so going for `--platform=all` will yield `darwin`, `linux` and `win32`. And if wanting another `mas` build together as well, they could do `--platform=mas`?

**Comment by malept** [comment]
> Just added in the rest of the documentation, and separated your commit for the basic Mac App Store docs in this PR.

The PR says there are still merge conflicts (and the GitHub diff is very messy). I suggest:

1. check out https://github.com/malept/electron-packager/tree/mas-squashed
2. add your commit to add additional documentation
3. force push onto https://github.com/sethlu/electron-packager/tree/master

**Comment by sethlu** [comment]
@malept Just found that I previously rebased on your branch. Now it's done again on `maxogden/electron-packager master`. (Now it turns out that Travis comes back again, which is great.)
I'll have a check with the code and readme/usage notes first and then `git cherry-pick` your added docs.

**Comment by sethlu** [comment]
@malept The diffs except `test/mac.js` should be quite neat. As I have to reduce the amount of testing script written for `darwin` and `mas`, the `test/mac.js` thus becomes a script shared by `test/darwin.js` and `test/mas.js`, each line with an indentation starting from the first testing function.
Also, the `NSHumanReadableCopyright` had earlier been added, with `--app-copyright` as the option, which fixes part of https://github.com/maxogden/electron-packager/issues/163 (1).

**Comment by malept** [comment]
The docs don't match what the code does: the code says the option name is `app-human-readable-copyright` while the docs say it's `app-copyright`.

Could you please refrain from adding more features to this PR? It's getting pretty difficult to see what your new changes are. I've also been holding off on merging other PRs because this one touches so much code.

**Comment by malept** [comment]
> It's getting pretty difficult to see what your new changes are.

This is also why in my rebase/squash branch of your PR, I split out your MAS changes from your "filter special characters in bundle ID" changes.

**Comment by sethlu** [comment]
@malept Just separated the commits to make each change clearer. Mind having another check?

**Comment by malept** [comment]
Once the `app-copyright` docs are updated, I'm going to merge this. (Unless one of the other contributors raises concerns soon - @kfranqueiro's concern about whether adding this feature constitutes a major or minor version bump will be addressed in #266. Hopefully I didn't miss anything.)

**Comment by sethlu** [comment]
@malept The `readme.md` has been updated with `app-copyright`.
I'm not sure but there's really quite a lot of working in changing the plist properties from my commits.

File: readme.md [hunk 10]
```diff
@@ -137,6 +153,10 @@ packager(opts, function done (err, appPath) { })
 
   Valid values are listed in [Apple's documentation](https://developer.apple.com/library/ios/documentation/General/Reference/InfoPlistKeyReference/Articles/LaunchServicesKeys.html#//apple_ref/doc/uid/TP40009250-SW8).
 
+`app-copyright` - *String*
+
+  The copyrights string to use in the app plist, will be displayed in the application About box (OS X only).
```
Comment (malept) [inline readme.md]: "copyright", not "copyrights string". I'll fix this post-merge. (I'm doing #265 anyway.)

**Comment by malept** [comment]
Excellent, I'm going to merge this later today.

> I'm not sure but there's really quite a lot of working in changing the plist properties from my commits.

Yes, @maxogden had a similar observation in a different PR: https://github.com/maxogden/electron-packager/pull/253#issuecomment-180172918

**Comment by sethlu** [comment]
@malept cool! After this I could join their discussions on plists, just to minimize the amount of code based on file modifications.

**Comment by malept** [comment]
Oops, forgot to merge this last night. I will do so now.

Thanks @sethlu for all the work you've done on this pull request!

**Comment by jasonhinkle** [comment]
Fantastic work guys!

**Comment by positlabs** [comment]
@sethlu @malept, if you're ever in Oakland, CA, I owe you a beer!

**Pull request merged**